src/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncEngine.Unix.cs

```diff
@@ -127,13 +140,22 @@ private static void AllocateToken(SocketAsyncContext context, out SocketAsyncEng
         {
             lock (s_lock)
             {
-                if (s_currentEngine == null)
+                engine = s_currentEngines[s_allocateFromEngine];
+                if (engine == null)
                 {
-                    s_currentEngine = new SocketAsyncEngine();
+                    s_currentEngines[s_allocateFromEngine] = engine = new SocketAsyncEngine();
                 }
 
-                engine = s_currentEngine;
-                handle = s_currentEngine.AllocateHandle(context);
+                handle = engine.AllocateHandle(context);
+
+                if (engine.IsFull)
+                {
+                    // We'll need to create a new event port for the next handle.
+                    s_currentEngines[s_allocateFromEngine] = null;
+                }
+
+                // Round-robin to the next engine
```

Comment:
Indeed, for a small number of sockets this is not optimal.
I optimized for the server case and tried to limit the impact on small systems via the ProcessorCount >= 6 condition.

Sockets could be assigned to engines based on load, but then for each allocation you need to iterate over the socket engines to figure out the least loaded. Round-robin should do a fairly good job of distributing the load.

It may be interesting to defer the creation of additional engines until the nr of active sockets on the first reaches a certain threshold.
Unless there is load-based assignment, a small threshold may be preferable to avoid the initial un-even distribution.